Userland/Applications/PixelPaint/Tools/BrushTool.cpp

```diff
@@ -199,16 +220,32 @@ NonnullRefPtr<Gfx::Bitmap> BrushTool::build_cursor()
     painter.draw_line({ scaled_size, scaled_size - 5 }, { scaled_size, scaled_size + 5 }, Color::LightGray, 3);
     painter.draw_line({ scaled_size - 5, scaled_size }, { scaled_size + 5, scaled_size }, Color::MidGray, 1);
     painter.draw_line({ scaled_size, scaled_size - 5 }, { scaled_size, scaled_size + 5 }, Color::MidGray, 1);
-    aa_painter.draw_ellipse(Gfx::IntRect(0, 0, containing_box_size, containing_box_size), Color::LightGray, 1);
+
+    // If no ImageEditor is present, we cannot bind the ellipse within the editor. Then we will just draw the ellipse.
+    if (!m_editor) {
+        aa_painter.draw_ellipse(Gfx::IntRect(0, 0, containing_box_size, containing_box_size), Color::LightGray, 1);
+    } else if (m_current_tool_position.x() + scaled_size <= m_editor->width() && m_current_tool_position.y() + scaled_size <= m_editor->height() && m_current_tool_position.x() - scaled_size > 0 && m_current_tool_position.y() - scaled_size > 0) {
+        Color color = m_editor->color_for(GUI::MouseButton::Primary);
+        aa_painter.fill_ellipse(Gfx::IntRect(0, 0, containing_box_size, containing_box_size), color.with_alpha(100));
+    }
+
+    m_scale_last_created_cursor = new_scale;
 
     return new_cursor;
 }
 
 void BrushTool::refresh_editor_cursor()
 {
+    // FIXME: reverting the change to only build_cursor on m_editor scale size or tool size causes slow behavior when zoomed to max with the BrushTool
```

Comment:
When this function get's called on every mouse move that the editor recognizes it currently always seem to allocate and build a new cursor bitmap. That should be avoided if possible.

Can't the current position been handled by the tool itself instead of rely on the "outer" world to manage it correctly?